osquery/sql/sqlite_filesystem.cpp

```diff
@@ -0,0 +1,166 @@
+/**
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <boost/filesystem.hpp>
+#include <boost/optional.hpp>
+
+#include <sqlite3.h>
+
+namespace osquery {
+  
+static boost::optional<std::string> findExistingProgramPathFromCommand(
+    const char *path, char escape_symbol, bool allow_quoting, bool shortest) {
+  size_t length = strlen(path);
+  std::string result;
+  size_t pos = 0;
+  // Skip spaces
+  for (; pos < length; ++pos) {
+    if (!isspace(path[pos])) {
+      break;
+    }
+  }
+  std::string temp_string;
+  bool is_quoted = false;
+  bool is_escaped = false;
+  for (; pos < length; ++pos) {
+    if (is_escaped) {
+      temp_string += path[pos];
+      is_escaped = false;
+      continue;
+    }
+    if (allow_quoting && path[pos] == '"') {
+      is_quoted = !is_quoted;
+      continue;
+    }
+    if (path[pos] == escape_symbol) {
+      is_escaped = true;
+      continue;
+    }
+    if (!is_quoted && isspace(path[pos])) {
+      // validate temp string
+      boost::filesystem::path test_path = temp_string;
+      auto status = boost::filesystem::status(test_path);
+      if (boost::filesystem::exists(status) &&
+          !boost::filesystem::is_directory(status)) {
+        result = temp_string;
+        if (shortest) {
+          break;
+        }
+      }
+    }
+    temp_string += path[pos];
+  }
+  if (result.length() == 0 || !shortest) {
+    boost::filesystem::path test_path = temp_string;
+    auto status = boost::filesystem::status(test_path);
+    if (boost::filesystem::exists(status) &&
+        !boost::filesystem::is_directory(status)) {
+      result = temp_string;
+    }
+  }
+  return result;
+}
+
+static boost::optional<std::string> findExistingProgramPathFromCommandSqlArgs(
+    int argc, sqlite3_value** argv, bool shortest) {
+  if (argc == 0) {
+    return boost::none;
+  }
+  // NULLs are not allowed
+  for (int i = 0; i < argc; i++) {
+    if (SQLITE_NULL == sqlite3_value_type(argv[i])) {
+      return boost::none;
+    }
+  }
+  const char* path = (const char*)sqlite3_value_text(argv[0]);
+  bool allow_quoting = false;
+  if (argc > 1) {
+    allow_quoting = sqlite3_value_int(argv[1]) != 0 ? true : false;
+  }
+#ifdef WIN32
```

Comment:
We have some helper functions to differentiate platform as we prefer to avoid preprocessor defines whenever possible. Would you mind making this:
```
if (isPlatform(PlatformType::TYPE_WINDOWS) {
  ...
} else {
  ...
}
```

Reply:
Better yet you could likely make this a ternary statement:
```
auto escape_symbol = isPlatform(PlatformType::TYPE_WINDOWS) ? '\\' : '^';
```

Reply:
@mkareta any reason you didn't update this? I think this is the only thing I see that would still need fixed.

Reply:
What is the reason to use run time check vs compile time, I understand why we need runtime check for tables and related stuff, but here I think is just small performance win to not do if's